File: python/integracion_con_bases_de_datos/mini_sistema_completo/mini_sistema_completo.py
from conexion import conectar
import re
def validar_nombre(nombre):
    return len(nombre.strip()) > 0
def validar_telefono(telefono):
    return telefono.isdigit() and len(telefono) >= 7
def validar_email(email):
    patron = r'^[\w\.-]+@[\w\.-]+\.\w+$'
    return re.match(patron, email)
def guardar_contacto(nombre, telefono, email):
    if not validar_nombre(nombre):
        print("❌ Nombre inválido")
        return
    if not validar_telefono(telefono):
        print("❌ Teléfono inválido")
        return
    if not validar_email(email):
        print("❌ Email inválido")
        return
    conexion = conectar()
    cursor = conexion.cursor()
    try:
        sql = "INSERT INTO contactos (nombre, telefono, email) VALUES (%s, %s, %s)"
        cursor.execute(sql, (nombre, telefono, email))
        conexion.commit()
        print("✅ Contacto guardado")
    except Exception as e:
        print("⚠️ Error:", e)
    finally:
        cursor.close()
        conexion.close()
def mostrar_contactos():
    conexion = conectar()
    cursor = conexion.cursor()
    cursor.execute("SELECT * FROM contactos")
    resultados = cursor.fetchall()
    if resultados:
        print("\n📋 Contactos:\n")
        for c in resultados:
            print(f"ID: {c[0]} | Nombre: {c[1]} | Tel: {c[2]} | Email: {c[3]}")
    else:
        print("⚠️ No hay contactos")
    cursor.close()
    conexion.close()
def buscar_contacto(nombre):
    conexion = conectar()
    cursor = conexion.cursor()
    sql = "SELECT * FROM contactos WHERE nombre LIKE %s"
    cursor.execute(sql, (f"%{nombre}%",))
    resultados = cursor.fetchall()
    if resultados:
        print("\n🔍 Resultados:\n")
        for c in resultados:
            print(f"ID: {c[0]} | Nombre: {c[1]} | Tel: {c[2]} | Email: {c[3]}")
    else:
        print("⚠️ No se encontraron coincidencias")
    cursor.close()
    conexion.close()
def eliminar_contacto(id_contacto):
    conexion = conectar()
    cursor = conexion.cursor()
    cursor.execute("DELETE FROM contactos WHERE id = %s", (id_contacto,))
    conexion.commit()
    if cursor.rowcount > 0:
        print("🗑️ Contacto eliminado")
    else:
        print("⚠️ No encontrado")
    cursor.close()
    conexion.close()
def actualizar_contacto(id_contacto, nombre, telefono, email):
    conexion = conectar()
    cursor = conexion.cursor()
    sql = """
    UPDATE contactos 
    SET nombre=%s, telefono=%s, email=%s 
    WHERE id=%s
    """
    cursor.execute(sql, (nombre, telefono, email, id_contacto))
    conexion.commit()
    if cursor.rowcount > 0:
        print("✏️ Contacto actualizado")
    else:
        print("⚠️ No encontrado")
    cursor.close()
    conexion.close()
def ordenar_contactos(tipo):
    conexion = conectar()
    cursor = conexion.cursor()
    if tipo == "1":
        sql = "SELECT * FROM contactos ORDER BY nombre ASC"
    else:
        sql = "SELECT * FROM contactos ORDER BY nombre DESC"
    cursor.execute(sql)
    resultados = cursor.fetchall()
    for c in resultados:
        print(f"ID: {c[0]} | Nombre: {c[1]} | Tel: {c[2]} | Email: {c[3]}")
    cursor.close()
    conexion.close()
def main():
    while True:
        print("\n=== MINI SISTEMA DE AGENDA ===")
        print("1. Agregar contacto")
        print("2. Mostrar contactos")
        print("3. Buscar contacto")
        print("4. Actualizar contacto")
        print("5. Eliminar contacto")
        print("6. Ordenar A-Z")
        print("7. Ordenar Z-A")
        print("8. Salir")
        opcion = input("Seleccione: ")
        if opcion == "1":
            nombre = input("Nombre: ")
            telefono = input("Teléfono: ")
            email = input("Email: ")
            guardar_contacto(nombre, telefono, email)
        elif opcion == "2":
            mostrar_contactos()
        elif opcion == "3":
            nombre = input("Buscar nombre: ")
            buscar_contacto(nombre)
        elif opcion == "4":
            mostrar_contactos()
            idc = int(input("ID a actualizar: "))
            nombre = input("Nuevo nombre: ")
            telefono = input("Nuevo teléfono: ")
            email = input("Nuevo email: ")
            actualizar_contacto(idc, nombre, telefono, email)
        elif opcion == "5":
            mostrar_contactos()
            idc = int(input("ID a eliminar: "))
            eliminar_contacto(idc)
        elif opcion == "6":
            ordenar_contactos("1")
        elif opcion == "7":
            ordenar_contactos("2")
        elif opcion == "8":
            print("👋 Hasta luego")
            break
        else:
            print("❌ Opción inválida")
if __name__ == "__main__":
    main()

File: java/integracion_con_base_de_datos/mini_sistema_completo/src/mini_sistema_completo/mini_sistema_completo.java
package mini_sistema_completo;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.Scanner;
public class mini_sistema_completo {
    static Scanner sc = new Scanner(System.in);
    public static void main(String[] args) {
        Connection conn = conexion.conectar();
        if (conn == null) {
            System.out.println("❌ No se pudo conectar a la base de datos");
            return;
        }
        int opcion;
        do {
            System.out.println("\n=== MINI SISTEMA DE AGENDA ===");
            System.out.println("1. Agregar contacto");
            System.out.println("2. Mostrar contactos");
            System.out.println("3. Buscar contacto por nombre");
            System.out.println("4. Actualizar contacto");
            System.out.println("5. Eliminar contacto");
            System.out.println("6. Salir");
            System.out.print("Seleccione una opción: ");
            opcion = sc.nextInt();
            sc.nextLine();
            switch (opcion) {
                case 1:
                    agregarContacto(conn);
                    break;
                case 2:
                    mostrarContactos(conn);
                    break;
                case 3:
                    buscarContacto(conn);
                    break;
                case 4:
                    actualizarContacto(conn);
                    break;
                case 5:
                    eliminarContacto(conn);
                    break;
                case 6:
                    System.out.println("👋 Saliendo del sistema...");
                    break;
                default:
                    System.out.println("❌ Opción inválida");
            }
        } while (opcion != 6);
        sc.close();
    }
    public static void agregarContacto(Connection conn) {
        try {
            System.out.println("\n=== AGREGAR CONTACTO ===");
            System.out.print("Nombre: ");
            String nombre = sc.nextLine();
            System.out.print("Teléfono: ");
            String telefono = sc.nextLine();
            System.out.print("Correo: ");
            String correo = sc.nextLine();
            System.out.print("Dirección: ");
            String direccion = sc.nextLine();
            System.out.print("Categoría: ");
            String categoria = sc.nextLine();
            String sql = "INSERT INTO contactos (nombre, telefono, correo, direccion, categoria) VALUES (?, ?, ?, ?, ?)";
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setString(1, nombre);
            ps.setString(2, telefono);
            ps.setString(3, correo);
            ps.setString(4, direccion);
            ps.setString(5, categoria);
            ps.executeUpdate();
            System.out.println("✅ Contacto agregado correctamente");
        } catch (Exception e) {
            System.out.println("❌ Error al agregar contacto: " + e.getMessage());
        }
    }
    public static void mostrarContactos(Connection conn) {
        try {
            String sql = "SELECT * FROM contactos";
            PreparedStatement ps = conn.prepareStatement(sql);
            ResultSet rs = ps.executeQuery();
            System.out.println("\n=== LISTA DE CONTACTOS ===");
            while (rs.next()) {
                System.out.println("ID: " + rs.getInt("id"));
                System.out.println("Nombre: " + rs.getString("nombre"));
                System.out.println("Teléfono: " + rs.getString("telefono"));
                System.out.println("Correo: " + rs.getString("correo"));
                System.out.println("Dirección: " + rs.getString("direccion"));
                System.out.println("Categoría: " + rs.getString("categoria"));
                System.out.println("---------------------------");
            }
        } catch (Exception e) {
            System.out.println("❌ Error al mostrar contactos: " + e.getMessage());
        }
    }
    public static void buscarContacto(Connection conn) {
        try {
            System.out.print("Ingrese el nombre a buscar: ");
            String nombre = sc.nextLine();
            String sql = "SELECT * FROM contactos WHERE nombre LIKE ?";
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setString(1, "%" + nombre + "%");
            ResultSet rs = ps.executeQuery();
            while (rs.next()) {
                System.out.println("ID: " + rs.getInt("id"));
                System.out.println("Nombre: " + rs.getString("nombre"));
                System.out.println("Teléfono: " + rs.getString("telefono"));
                System.out.println("Correo: " + rs.getString("correo"));
                System.out.println("Dirección: " + rs.getString("direccion"));
                System.out.println("Categoría: " + rs.getString("categoria"));
                System.out.println("---------------------------");
            }
        } catch (Exception e) {
            System.out.println("❌ Error en la búsqueda: " + e.getMessage());
        }
    }
    public static void actualizarContacto(Connection conn) {
        try {
            System.out.print("Ingrese el ID del contacto: ");
            int id = sc.nextInt();
            sc.nextLine();
            System.out.print("Nuevo nombre: ");
            String nombre = sc.nextLine();
            System.out.print("Nuevo teléfono: ");
            String telefono = sc.nextLine();
            System.out.print("Nuevo correo: ");
            String correo = sc.nextLine();
            System.out.print("Nueva dirección: ");
            String direccion = sc.nextLine();
            System.out.print("Nueva categoría: ");
            String categoria = sc.nextLine();
            String sql = "UPDATE contactos SET nombre=?, telefono=?, correo=?, direccion=?, categoria=? WHERE id=?";
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setString(1, nombre);
            ps.setString(2, telefono);
            ps.setString(3, correo);
            ps.setString(4, direccion);
            ps.setString(5, categoria);
            ps.setInt(6, id);
            ps.executeUpdate();
            System.out.println("✅ Contacto actualizado");
        } catch (Exception e) {
            System.out.println("❌ Error al actualizar: " + e.getMessage());
        }
    }
    public static void eliminarContacto(Connection conn) {
        try {
            System.out.print("Ingrese el ID del contacto a eliminar: ");
            int id = sc.nextInt();
            String sql = "DELETE FROM contactos WHERE id=?";
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setInt(1, id);
            ps.executeUpdate();
            System.out.println("✅ Contacto eliminado");
        } catch (Exception e) {
            System.out.println("❌ Error al eliminar: " + e.getMessage());
        }
    }
}